velox/vector/tests/VectorTest.cpp

```diff
@@ -3535,5 +3535,84 @@ TEST_F(VectorTest, hashAll) {
   }
 }
 
+TEST_F(VectorTest, setType) {
+  auto type = ROW({"aa"}, {BIGINT()});
+  auto vector = BaseVector::create(type, 1'000, pool());
+
+  auto newType = ROW({"bb"}, {BIGINT()});
+  vector->setType(newType);
+
+  EXPECT_EQ(
+      std::static_pointer_cast<const RowType>(vector->type())->names()[0],
+      "bb");
+
+  auto failedType = ROW({"bb"}, {VARCHAR()});
+  VELOX_ASSERT_RUNTIME_THROW(
+      vector->setType(failedType),
+      "Cannot change vector type from ROW<bb:BIGINT> to ROW<bb:VARCHAR>. The old and new types can be different logical types, but the underlying physical types must match.")
+}
+
+TEST_F(VectorTest, setROWNestROWType) {
+  auto type =
+      ROW({"a", "b"}, {ROW({"c", "d"}, {BIGINT(), BIGINT()}), BIGINT()});
+  auto vector = BaseVector::create(type, 1'000, pool());
+
+  auto newType =
+      ROW({"a", "b"}, {ROW({"cc", "dd"}, {BIGINT(), BIGINT()}), BIGINT()});
+  vector->setType(newType);
+
+  auto rowType = std::static_pointer_cast<const RowType>(vector->type());
```

Comment:
There is a lot of code repetition in all these tests. Consider consolidating into a single test and share testing logic.

```
auto test = [&] (type, newType, invalidNewType) {
    auto vector = BaseVector::create(type, 1'000, pool());

    vector->setType(newType);
    EXPECT_EQ(vector->type()->toString(), newType->toString());    
    
    VELOX_ASSERT_RUNTIME_THROW(vector->setType(invalidNewType),...)
}

test(...);
test(...);
...

```